src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs

```diff
@@ -9,12 +9,27 @@
 namespace System
 {
     /// <summary>
+    /// This interface is required for types that want to be indexed into by dynamic patterns.
+    /// </summary>
+    public interface ITuple
+    {
+        /// <summary>
+        /// The number of positions in this data structure.
+        /// </summary>
+        int Length { get; }
+
+        /// <summary>
+        /// Get the element at position <param name="index"/>.
+        /// </summary>
+        object this[int index] { get; }
+    }
```

Comment:
@terrajobst From discussion with @gafter, he clarified that the `ITuple` interface would most likely only need to be implemented by `System.Tuple` and probably no other types. 
Also, it can be an explicit implementation, avoiding adding members to the public API.